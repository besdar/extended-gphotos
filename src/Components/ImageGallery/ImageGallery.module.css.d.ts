declare const styles: {
  readonly "gallery": string;
};
export = styles;

