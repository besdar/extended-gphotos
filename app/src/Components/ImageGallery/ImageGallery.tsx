import React, { FC, useEffect, useState } from 'react';
import ImageItem, { ImageItemPropsType, ImageItemType } from '../ImageItem/ImageItem';
import s from './ImageGallery.module.css';
import { getGalleryStructureData } from '../GalleryStructure';

const changeImageInArray = (
  ImageArray: Array<ImageItemPropsType & { id: number }>,
  getRandomImage: getRandomImageFuncType,
  simultaneouslyImageChanging = 3,
) => {
  const ImagesData = [...ImageArray];

  if (ImagesData.length !== 0) {
    const uniqueIndexes = new Set<number>();
    for (let i = 0; i < simultaneouslyImageChanging; i += 1) {
      uniqueIndexes.add(Math.floor(Math.random() * (ImagesData.length - 1)));
    }

    uniqueIndexes.forEach((randomIndex) => {
      const randomImage = getRandomImage();

      if (ImagesData[randomIndex].isFirstImageActive) {
        ImagesData[randomIndex] = {
          ...ImagesData[randomIndex],
          isFirstImageActive: false,
          anotherImageUrl: randomImage.imageUrl,
          tooltipTitle: randomImage.tooltipTitle,
          webUrl: randomImage.webUrl,
        };
      } else {
        ImagesData[randomIndex] = {
          ...ImagesData[randomIndex],
          isFirstImageActive: true,
          imageUrl: randomImage.imageUrl,
          tooltipTitle: randomImage.tooltipTitle,
          webUrl: randomImage.webUrl,
        };
      }
    });
  }

  return ImagesData;
};

type getRandomImageFuncType = () => ImageItemType;
type galleryStructureParamsType = [
        mapHeight: number,
        mapWidth: number,
        minSquareSide: number,
        timesNumberMaxSideGreaterThanMinSide: number,
    ];
type ImageGalleryPropsType = {
    simultaneouslyImageChanging?: number,
    getRandomImage: getRandomImageFuncType,
    galleryStructureConfig?: galleryStructureParamsType
}

const ImageGallery: FC<ImageGalleryPropsType> = ({
  getRandomImage,
  galleryStructureConfig,
  simultaneouslyImageChanging,
}) => {
  const [imagesData, setImagesData] = useState([] as Array<ImageItemPropsType & { id: number }>);
  const [rowLength, setRowLength] = useState(1);
  const congigHeight = galleryStructureConfig && galleryStructureConfig[0];
  const containerHeight = congigHeight || window.innerHeight;

  useEffect(() => {
    const galleryStructureParams: galleryStructureParamsType = galleryStructureConfig || [
      window.innerHeight,
      window.innerWidth,
      100,
      3,
    ];
    const {
      imagesStructureData,
      rowLength: imagesStructureRowLength,
    } = getGalleryStructureData(...galleryStructureParams);

    let counter = 0;
    const imageDataFromStructure = imagesStructureData.map((imageStructureData) => {
      const randomImageInfo = getRandomImage();
      counter += 1;

      return {
        squareCoordinates: {
          ...imageStructureData,
        },
        ...randomImageInfo,
        anotherImageUrl: randomImageInfo.imageUrl,
        isFirstImageActive: true,
        id: counter,
      };
    });

    setImagesData(imageDataFromStructure);
    setRowLength(imagesStructureRowLength);

    const interval = setInterval(() => {
      setImagesData((imagesDataOld) => (
        changeImageInArray(imagesDataOld, getRandomImage, simultaneouslyImageChanging)
      ));
    }, 4000);

    return () => clearInterval(interval);
  }, [galleryStructureConfig, getRandomImage, simultaneouslyImageChanging]);

  return (
    <div
      className={s.gallery}
      style={{ height: containerHeight, gridTemplateColumns: `repeat(${rowLength}, 1fr)` }}
    >
      {imagesData.map((imageProps) => (
        <ImageItem
          key={imageProps.id}
          anotherImageUrl={imageProps.anotherImageUrl}
          isFirstImageActive={imageProps.isFirstImageActive}
          squareCoordinates={imageProps.squareCoordinates}
          imageUrl={imageProps.imageUrl}
          tooltipTitle={imageProps.tooltipTitle}
          webUrl={imageProps.webUrl}
        />
      ))}
    </div>
  );
};

export default ImageGallery;
