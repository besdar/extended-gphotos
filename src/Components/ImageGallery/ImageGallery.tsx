import React, { FC, useEffect, useState } from 'react';
import ImageItem, { ImageItemPropsType, ImageItemType } from '../ImageItem/ImageItem';
import * as s from './ImageGallery.module.css';
import { galleryStructureParamsType, getGalleryStructureData } from '../GalleryStructure';

const changeImageInArray = (
  ImageArray: Array<ImagesDataType>,
  getRandomImage: getRandomImageFuncType,
  simultaneouslyImageChanging: number,
) => {
  if (
    ImageArray.length === 0
    || simultaneouslyImageChanging <= 0
    || simultaneouslyImageChanging >= ImageArray.length - 1
  ) {
    return ImageArray;
  }

  const ImagesData = [...ImageArray];

  const uniqueIndexes = new Set<number>();
  while (uniqueIndexes.size !== simultaneouslyImageChanging) {
    uniqueIndexes.add(Math.floor(Math.random() * (ImagesData.length - 1)));
  }

  uniqueIndexes.forEach((randomIndex) => {
    const randomImage = getRandomImage();

    if (ImagesData[randomIndex].isFirstImageActive) {
      ImagesData[randomIndex] = {
        ...ImagesData[randomIndex],
        isFirstImageActive: false,
        anotherImageUrl: randomImage.imageUrl,
        tooltipTitle: randomImage.tooltipTitle,
        webUrl: randomImage.webUrl,
      };
    } else {
      ImagesData[randomIndex] = {
        ...ImagesData[randomIndex],
        isFirstImageActive: true,
        imageUrl: randomImage.imageUrl,
        tooltipTitle: randomImage.tooltipTitle,
        webUrl: randomImage.webUrl,
      };
    }
  });

  return ImagesData;
};

type ImagesDataType = ImageItemPropsType & { id: number };
type getRandomImageFuncType = () => ImageItemType;
type ImageGalleryPropsType = {
  getRandomImage: getRandomImageFuncType,
  galleryStructureConfig: galleryStructureParamsType,
  containerHeight?: number | string,
  simultaneouslyImageChanging?: number,
}

const ImageGallery: FC<ImageGalleryPropsType> = ({
  getRandomImage,
  galleryStructureConfig,
  containerHeight = '100%',
  simultaneouslyImageChanging = 3,
}) => {
  const [imagesData, setImagesData] = useState([] as Array<ImagesDataType>);

  useEffect(() => {
    const imagesStructureData = getGalleryStructureData(galleryStructureConfig);

    let counter = 0;
    const imageDataFromStructure = imagesStructureData.map((imageStructureData) => {
      const randomImageInfo = getRandomImage();
      counter += 1;

      return {
        squareCoordinates: {
          ...imageStructureData,
        },
        ...randomImageInfo,
        anotherImageUrl: randomImageInfo.imageUrl,
        isFirstImageActive: true,
        id: counter,
      };
    });

    setImagesData(imageDataFromStructure);

    const interval = setInterval(() => {
      setImagesData((imagesDataOld) => (
        changeImageInArray(imagesDataOld, getRandomImage, simultaneouslyImageChanging)
      ));
    }, 4000);

    return () => clearInterval(interval);
  }, [galleryStructureConfig, getRandomImage, simultaneouslyImageChanging]);

  return (
    <div
      className={s.gallery}
      style={{ height: containerHeight, gridTemplateColumns: `repeat(${galleryStructureConfig.columnsCount}, 1fr)` }}
    >
      {imagesData.map((imageProps) => (
        <ImageItem
          key={imageProps.id}
          anotherImageUrl={imageProps.anotherImageUrl}
          isFirstImageActive={imageProps.isFirstImageActive}
          squareCoordinates={imageProps.squareCoordinates}
          imageUrl={imageProps.imageUrl}
          tooltipTitle={imageProps.tooltipTitle}
          webUrl={imageProps.webUrl}
        />
      ))}
    </div>
  );
};

export default ImageGallery;
