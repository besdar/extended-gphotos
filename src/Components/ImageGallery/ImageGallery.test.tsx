import { mount } from 'enzyme';
import React from 'react';
import ImageGallery from './ImageGallery';
import { act } from 'react-dom/test-utils';

it('Should normal first render with normal props values', async () => {
    jest.useFakeTimers();

    let component;
    await act(async () => {
        component = mount(
            <ImageGallery
                galleryStructureConfig={{
                    columnsCount: 3,
                    rowsCount: 3,
                    maxSquareSizeInColumns: 2
                }}
                getRandomImage={() => ({
                    imageUrl: 'imageUrl',
                    tooltipTitle: 'tooltipTitle',
                    webUrl: 'webUrl',
                })}
            />
        );

        jest.runOnlyPendingTimers();
    });

    expect(component).toMatchSnapshot();
});

it('Should normal render with first useEffect with normal props values', async () => {
    jest.useFakeTimers();

    let component;
    await act(async () => {
        component = mount(
            <ImageGallery
                galleryStructureConfig={{
                    columnsCount: 3,
                    rowsCount: 3,
                    maxSquareSizeInColumns: 2
                }}
                getRandomImage={() => ({
                    imageUrl: 'imageUrl',
                    tooltipTitle: 'tooltipTitle',
                    webUrl: 'webUrl',
                })}
            />
        );

        jest.runOnlyPendingTimers();
        jest.runOnlyPendingTimers();
    });

    expect(component).toMatchSnapshot();
});

it('Should normal render with useEffect update with normal props values', async () => {
    jest.useFakeTimers();

    let component;
    await act(async () => {
        component = mount(
            <ImageGallery
                galleryStructureConfig={{
                    columnsCount: 3,
                    rowsCount: 3,
                    maxSquareSizeInColumns: 2
                }}
                getRandomImage={() => ({
                    imageUrl: 'imageUrl',
                    tooltipTitle: 'tooltipTitle',
                    webUrl: 'webUrl',
                })}
            />
        );

        jest.runOnlyPendingTimers();
        jest.runOnlyPendingTimers();
        jest.runOnlyPendingTimers();
    });

    expect(component).toMatchSnapshot();
});

it('Should not rerender render with useEffect update with wrong props values', async () => {
    jest.useFakeTimers();

    let component;
    await act(async () => {
        component = mount(
            <ImageGallery
                galleryStructureConfig={{
                    columnsCount: 3,
                    rowsCount: 3,
                    maxSquareSizeInColumns: 2
                }}
                getRandomImage={() => ({
                    imageUrl: 'imageUrl',
                    tooltipTitle: 'tooltipTitle',
                    webUrl: 'webUrl',
                })}
                simultaneouslyImageChanging={0}
            />
        );

        jest.runOnlyPendingTimers();
        jest.runOnlyPendingTimers();
    });

    expect(component).toMatchSnapshot();
});