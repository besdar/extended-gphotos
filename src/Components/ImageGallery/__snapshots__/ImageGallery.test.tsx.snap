// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should normal first render with normal props values 1`] = `ReactWrapper {}`;

exports[`Should normal render with first useEffect with normal props values 1`] = `ReactWrapper {}`;

exports[`Should normal render with useEffect update with normal props values 1`] = `ReactWrapper {}`;

exports[`Should not rerender render with useEffect update with wrong props values 1`] = `ReactWrapper {}`;
